gotham/src/handler/static_file.rs

```diff
@@ -0,0 +1,246 @@
+use http::response::{create_response, extend_response};
+use router::response::extender::StaticResponseExtender;
+use state::{FromState, State, StateData};
+use hyper;
+use mime::{self, Mime};
+use mime_guess::guess_mime_type_opt;
+use std::fs;
+use std::io::{self, Read};
+use std::path::{Component, Path, PathBuf};
+use std::iter::FromIterator;
+
+use futures::future;
+use handler::{Handler, HandlerFuture, NewHandler};
+
+/// Represents a handler for any files under the path `root`.
+#[derive(Clone)]
+pub struct FileSystemHandler {
+    root: PathBuf,
+}
+
+/// Represents a handler for a single file at `path`.
+#[derive(Clone)]
+pub struct FileHandler {
+    path: PathBuf,
+}
+
+impl FileHandler {
+    /// Create a new `FileHandler` for the given path.
+    pub fn new(path: &str) -> FileHandler {
+        FileHandler {
+            path: PathBuf::from(path),
+        }
+    }
+}
+
+impl FileSystemHandler {
+    /// Create a new `FileSystemHandler` with the given root path.
+    pub fn new(root: &str) -> FileSystemHandler {
+        FileSystemHandler {
+            root: PathBuf::from(root),
+        }
+    }
+}
+
+impl NewHandler for FileHandler {
+    type Instance = Self;
+
+    fn new_handler(&self) -> io::Result<Self::Instance> {
+        Ok(self.clone())
+    }
+}
+
+impl NewHandler for FileSystemHandler {
+    type Instance = Self;
+
+    fn new_handler(&self) -> io::Result<Self::Instance> {
+        Ok(self.clone())
+    }
+}
+
+impl Handler for FileSystemHandler {
+    fn handle(self, state: State) -> Box<HandlerFuture> {
+        let path = {
+            let mut base_path = PathBuf::from(self.root);
+            let file_path = PathBuf::from_iter(&FilePathExtractor::borrow_from(&state).parts);
+            base_path.extend(&normalize_path(&file_path));
+            base_path
+        };
+        let response = create_file_response(path, &state);
+        Box::new(future::ok((state, response)))
+    }
+}
+
+impl Handler for FileHandler {
+    fn handle(self, state: State) -> Box<HandlerFuture> {
+        let response = create_file_response(self.path, &state);
+        Box::new(future::ok((state, response)))
+    }
+}
+
+fn create_file_response(path: PathBuf, state: &State) -> hyper::Response {
```

Comment:
It might be cool to move this to a pub fn in response, so users could call it with their custom logic.

Reply:
Great suggestion, thanks. I've logged this as #206 so we don't hold up this PR discussing the API (which was my first instinct 😅).